File: sdk/go/service_registry.go
package sdk

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"regexp"

	"github.com/hyperledger/fabric-gateway/pkg/client"
	"github.com/nexus-lab/iot-service-blockchain/common"
)

// ServiceEvent an event emitted by the service registry contract notifying a service update
type ServiceEvent struct {
	// Action name of the action performed on the service
	Action string

	// OrganizationId organization ID of the service
	OrganizationId string

	// DeviceId ID of the device to which the service belongs
	DeviceId string

	// ServiceName name of the service
	ServiceName string

	// Payload custom event payload
	Payload interface{}
}

// ServiceRegistryInterface core utilities for managing services on the ledger
type ServiceRegistryInterface interface {
	// Register create or update a service in the ledger
	Register(service *common.Service) error

	// Get return a service by its organization ID, device ID, and name
	Get(organizationId string, deviceId string, serviceName string) (*common.Service, error)

	// GetAll return a list of services by their organization ID and device ID
	GetAll(organizationId string, deviceId string) ([]*common.Service, error)

	// Deregister remove a service from the ledger
	Deregister(service *common.Service) error

	// RegisterEvent registers for service registry events
	RegisterEvent(options ...client.ChaincodeEventsOption) (<-chan *ServiceEvent, context.CancelFunc, error)
}

// ServiceRegistry core utilities for managing services on the ledger
type ServiceRegistry struct {
	contract ContractInterface
}

// Register create or update a service in the ledger
func (r *ServiceRegistry) Register(service *common.Service) error {
	if service == nil {
		return fmt.Errorf("cannot register an empty service")
	}

	data, err := service.Serialize()
	if err != nil {
		return err
	}

	_, err = r.contract.SubmitTransaction("Register", string(data))
	return err
}

// Get return a service by its organization ID, device ID, and name
func (r *ServiceRegistry) Get(organizationId string, deviceId string, serviceName string) (*common.Service, error) {
	data, err := r.contract.SubmitTransaction("Get", organizationId, deviceId, serviceName)
	if err != nil {
		return nil, err
	}

	return common.DeserializeService(data)
}

// GetAll return a list of services by their organization ID and device ID
func (r *ServiceRegistry) GetAll(organizationId string, deviceId string) ([]*common.Service, error) {
	data, err := r.contract.SubmitTransaction("GetAll", organizationId, deviceId)
	if err != nil {
		return nil, err
	}

	results := make([]*common.Service, 0)
	if err = json.Unmarshal(data, &results); err != nil {
		return nil, err
	}

	return results, nil
}

// Deregister remove a service from the ledger
func (r *ServiceRegistry) Deregister(service *common.Service) error {
	if service == nil {
		return fmt.Errorf("cannot deregister an empty device")
	}

	data, err := service.Serialize()
	if err != nil {
		return err
	}

	_, err = r.contract.SubmitTransaction("Deregister", string(data))
	return err
}

// RegisterEvent registers for service registry events
func (r *ServiceRegistry) RegisterEvent(options ...client.ChaincodeEventsOption) (<-chan *ServiceEvent, context.CancelFunc, error) {
	dest := make(chan *ServiceEvent)
	source, cancel, err := r.contract.RegisterEvent(options...)
	pattern := regexp.MustCompile(`^service:\/\/(.+?)\/(.+?)\/(.+?)\/(.+?)$`)

	go func() {
		defer close(dest)

		for event := range source {
			matches := pattern.FindStringSubmatch(event.EventName)
			if len(matches) != 5 {
				continue
			}

			serviceEvent := &ServiceEvent{
				OrganizationId: matches[1],
				DeviceId:       matches[2],
				ServiceName:    matches[3],
				Action:         matches[4],
			}

			if serviceEvent.Action == "register" || serviceEvent.Action == "deregister" {
				service, err := common.DeserializeService(event.Payload)
				if err != nil {
					log.Printf("bad service event payload %#v, action is %s\n", event.Payload, serviceEvent.Action)
					continue
				}
				serviceEvent.Payload = service
			} else {
				serviceEvent.Payload = event.Payload
			}

			dest <- serviceEvent
		}
	}()

	return dest, cancel, err
}

func createServiceRegistry(network *client.Network, chaincodeId string) ServiceRegistryInterface {
	return &ServiceRegistry{
		contract: &Contract{
			network:      network,
			chaincodeId:  chaincodeId,
			contractName: "service_registry",
		},
	}
}

File: sdk/javascript/ServiceRegistry.ts
import {
  ChaincodeEventsOptions,
  CloseableAsyncIterable,
  Network,
} from '@hyperledger/fabric-gateway';
import { TextDecoder } from 'util';

import Contract, { ContractInterface } from './Contract';
import Service from './Service';

/**
 * Interface of the event emitted by the service registry contract notifying a service update
 */
export interface ServiceEvent {
  /**
   * Name of the action performed on the service
   */
  action: string;

  /**
   * Organization ID of the service
   */
  organizationId: string;

  /**
   * ID of the device to which the service belongs
   */
  deviceId: string;

  /**
   * Name of the service
   */
  serviceName: string;

  /**
   * Custom event payload
   */
  payload: any;
}

/**
 * Interface of core utilities for managing services on the ledger
 */
export interface ServiceRegistryInterface {
  /**
   * Create or update a service in the ledger
   *
   * @param service service to be created or updated
   */
  register(service: Service): Promise<void>;

  /**
   * Get a service by its organization ID, device ID, and name
   *
   * @param organizationId service's organization ID
   * @param deviceId device's ID
   * @param serviceName name of the service
   * @returns the device
   */
  get(organizationId: string, deviceId: string, serviceName: string): Promise<Service>;

  /**
   * Get a list of services by their organization ID and device ID
   *
   * @param organizationId device's organization ID
   * @param deviceId device's ID
   * @returns all services of the device in the organization
   */
  getAll(organizationId: string, deviceId: string): Promise<Service[]>;

  /**
   * Remove a service from the ledger
   *
   * @param service the service to be removed
   */
  deregister(device: Service): Promise<void>;

  /**
   * Registers for service registry events
   *
   * @param options chaincode event options
   * @returns an async iterable of the events
   */
  registerEvent(options?: ChaincodeEventsOptions): Promise<CloseableAsyncIterable<ServiceEvent>>;
}

/**
 * Core utilities for managing devices on the ledger
 */
export default class ServiceRegistry implements ServiceRegistryInterface {
  private utf8Decoder: TextDecoder;
  /**
   * @param contract smart contract
   */
  constructor(private contract: ContractInterface) {
    this.utf8Decoder = new TextDecoder();
  }

  async register(service: Service): Promise<void> {
    const serialized = service.serialize();
    await this.contract.submitTransaction('Register', serialized);
  }

  async get(organizationId: string, deviceId: string, serviceName: string): Promise<Service> {
    const data = await this.contract.submitTransaction(
      'Get',
      organizationId,
      deviceId,
      serviceName,
    );
    const serialized = this.utf8Decoder.decode(data);
    return Service.deserialize(serialized);
  }

  async getAll(organizationId: string, deviceId: string): Promise<Service[]> {
    const data = await this.contract.submitTransaction('GetAll', organizationId, deviceId);
    const serialized = this.utf8Decoder.decode(data);
    return (JSON.parse(serialized) as any[]).map(Service.fromObject);
  }

  async deregister(service: Service): Promise<void> {
    const serialized = service.serialize();
    await this.contract.submitTransaction('Deregister', serialized);
  }

  async registerEvent(
    options?: ChaincodeEventsOptions,
  ): Promise<CloseableAsyncIterable<ServiceEvent>> {
    const events = await this.contract.registerEvent(options);
    const pattern = /^service:\/\/(.+?)\/(.+?)\/(.+?)\/(.+?)$/;
    const decoder = this.utf8Decoder;

    return {
      async *[Symbol.asyncIterator]() {
        for await (const event of events) {
          // can reuse pattern here since it has no global('g') flag
          const matches = pattern.exec(event.eventName);
          if (matches === null || matches.length != 5) {
            continue;
          }

          const serviceEvent: ServiceEvent = {
            organizationId: matches[1],
            deviceId: matches[2],
            serviceName: matches[3],
            action: matches[4],
            payload: null,
          };

          if (serviceEvent.action === 'register' || serviceEvent.action === 'deregister') {
            try {
              const service = Service.deserialize(decoder.decode(event.payload));
              serviceEvent.payload = service;
            } catch {
              console.error(
                `bad service event payload ${event.payload}, action is ${serviceEvent.action}`,
              );
              continue;
            }
          } else {
            serviceEvent.payload = event.payload;
          }

          yield serviceEvent;
        }
      },
      close() {
        events.close();
      },
    };
  }
}

/**
 * The default factory for creating service registries
 * 
 * @param network Hyperledger Fabric network
 * @param chaincodeId ID/name of the chaincode
 * @returns The service registry
 */
export function createServiceRegistry(network: Network, chaincodeId: string) {
  return new ServiceRegistry(new Contract(network, chaincodeId, 'service_registry'));
}
